api-specificatie/openapi.yaml

```diff
@@ -33,26 +42,15 @@ paths:
           $ref: "https://raw.githubusercontent.com/VNG-Realisatie/Haal-Centraal-common/v1.0.0/api-specificatie/common.yaml#/components/responses/401"
         '403':
           $ref: "https://raw.githubusercontent.com/VNG-Realisatie/Haal-Centraal-common/v1.0.0/api-specificatie/common.yaml#/components/responses/403"
-        '409':
```

Comment:
409 is wel nodig. Namelijk wanneer een burgerservicenummer is opgenomen waarvoor al een volgindicatie bestaat. Je wilt denk ik voorkomen dat er bij dezelfde gebruiker meerdere volgindicaties bestaan op hetzelfde burgerservicenummer. Anders werkt DELETE /volgindicatiepersonen/{burgerservicenummer} niet meer.

Reply:
Ik vraag me af of je in de situatie die je beschrijft niet juist een 403 zou moeten gebruiken. De 409 is immers alleen toegestaan in situaties waarin je mag verwachten dat de gebruiker het conflict kan herstellen en een nieuw request kan sturen. Het sturen van een nieuw request is hier echter juist niet van toepassing. Bij een 403 wordt er vanuit gegaan dat de request niet opnieuw wordt ingediend.

Reply:
@melsk-r we gebruiken 403 voor autorisatie meldingen. Dat is denk ik hoe de meeste gebruikers 403 kennen. Ik denkt dat het verwarrend kan zijn 403 ook te gebruiken voor deze situatie. Bovendien kan de gebruiker het wel oplossen, bijvoorbeeld door eerst een DELETE te sturen.

Ik zie verschillende oplossingen:

1. 201 (geen foutmelding)
2. 400 Bad request
3. 403 Forbidden
4. 409 Conflict

Voordeel van de 201 is dat de gebruiker er geen "last" van heeft. Het resultaat is als verwacht, namelijk dat de volgindicatie bestaat. Dit is vooral handig wanneer meerdere volgindicaties (burgerservicenummer) tegelijk worden gestuurd, waarvan er slechts één of enkele (niet alle) al bestaan. Deze correcte (nieuwe) volgindicaties worden dan gewoon toegevoegd. De gebruiker hoeft geen nieuw request te sturen waar de burgerservicenummers die al bestaan uitgefilterd zijn.
Nadeel van de 201 is dat er eigenlijk een onjuist antwoord wordt gegeven, want de opgegeven volgindicatie is niet created (bestond immers al).

Voordeel van de 400 melding is dat dit past in de normale foutafhandeling die al geïmplementeerd wordt, en dat deze een invalidParams heeft waarin het betreffende bestaande burgerservicenummer expliciet gemeld kan worden. In een POST kunnen immers meerdere burgerservicenummers tegelijk worden opgegeven, waarvan een één of enkele (niet alle) al een volgindicatie hebben.
Nadeel van de 400 melding is dat deze eigenlijk niet juist is, het request heeft geen onjuiste structuur en is in die zin geen "bad request".

Voordeel van de 403 melding is dat deze duidelijk maakt dat dit request niet uitgevoerd mag worden ("forbidden") en opnieuw sturen geen zin heeft.
Nadeel is dat de 403 melding voor veel gebruikers geassocieerd wordt met autorisatie meldingen (de gebruiker heeft geen rechten op deze operatie).

Voordeel van de 409 melding is dat dit (na korte internet zoektocht) vaak wordt aanbevolen als de oplossing hiervoor.

Meldingen 403 en 409 kennen nu geen invalidParams component in de response. De gebruiker moet dus uit "detail" halen welke burgerservicenummers het probleem veroorzaken. Dit is een vrij tekstveld, dus dit kan alleen door een mens worden geïnterpreteerd. Geautomatiseerde afhandeling van deze situatie is dus niet mogelijk, althans niet wanneer er meer dan één burgerservicenummer is gestuurd in de POST.

@MelvLee en @KayodeBakker ik ben benieuwd hoe jullie hiernaar kijken. Wat is de meest gebruikersvriendelijke oplossing hiervoor?

Reply:
Als ik je betoog zo lees zou je zeggen dat de 201 nog niet zo'n slecht idee is. Je kunt je afvragen of er hier echt sprake is van een foutsituatie. Vergelijk het met de lichtschakelaar in een trappenhuis die er voor zorgt dat het licht een beperkte tijd aan gaat. Als je daar nog eens op drukt blijft het licht gewoon branden.

Dat verandert echter als we 'einddatum' alsnog aan de POST gaan toevoegen. Dan moet er wel degelijk iets worden gewijzigd en moet de POST ineens als een PATCH worden beschouwd. Maar op het moment dat we 'einddatum' alsnog gaan toevoegen moeten we waarschijnlijk sowieso meer zaken in de API gaan veranderen. Ik vermoed dat je dan om te kunnen bepalen of je een POST of PATCH moet sturen eerst een GET moet doen.

Reply:
Als we kiezen voor de 201 oplossing (negeren dat een volgindicatie al bestaat en een POST dus negeren wanneer deze al bestaat), mag dat niet veranderen bij toevoegen van einddatum (of enig ander gegeven). Dat zou een breaking change opleveren voor gebruikers. Ik denk dat de structuur nu zo is dat het toevoegen van gegevens kan zonder breaking change in het gebruik van de API voor gebruikers die alleen burgerservicenummers sturen.

De server zou in dat geval een POST moeten afhandelen als PUT (niet als PATCH!) wanneer er al een volgindicatie met dat burgerservicenummer bestaat.

Reply:
Ik ga er even vanuit dat we het over de post actie hebben.

Wellicht een functioneel vraagstuk dat al besproken is, maar waar ik niet bij was, maar waarom is er überhaupt een discussie voor een 409 of vergelijkbare oplossing?
De eerste keer dat Hans en ik de specificatie hadden gelezen, interpreteerde wij het als dat een abonnee **altijd** deze actie mag uitvoeren en als er al 1 bestaat dat deze dan wordt overschreven.

Bijv. eerst:

```json
{
  "volgindicatiepersonen": [
    {
      "burgerservicenummer": "555555021",
      "einddatum": "2020-08-10"
    }
  ]
}
```
en later blijkt dat om 1 of andere reden een verlenging nodig is.

```json
{
  "volgindicatiepersonen": [
    {
      "burgerservicenummer": "555555021",
      "einddatum": "2020-08-20"
    }
  ]
}
```

Is er een wettelijke of functionele reden waarom iemand de indicatie mag verlenen/eerder laten aflopen door nogmaals de post uit te voeren? Begrijpelijk wel is dat eerder laten aflopen weer tegenstrijdig is gezien we de DELETE actie hebben, maar het gaat meer even om de mogelijkheid tot verlenen waardoor ik me afvraag of we dit überhaupt willen tegenhouden. Eens dat je in geval dat je wil kunnen updaten voor verlening of dergelijke dat we dan PUT gebruiken zodat we niet een POST en PATCH hebben, maar 1 in dezelfde actie.

Reply:
@KayodeBakker  Neem in je overwegingen mee dat 'einddatum' zal worden verwijderd uit deze API (zie #19). Het wijzigen van de einddatum is dus vooralsnog (en wellicht wel nooit) im frage.

Reply:
Als je strict de REST constraints volgt, dan zou de POST van een bestaande 'resource' tot een fout moeten leiden, omdat binnen REST is gespecificeerd dat elke valide POST leidt tot een nieuwe 'resource'. Bij een PUT kan je er wel voor kiezen om de resource te creeren als deze nog niet bestaat.
Verder wordt in https://tools.ietf.org/html/rfc7231#section-6.5.8 aangegeven dat 409 meestal wordt gebruikt bij PUTs. Bij een POST van een bestaande resource zou ik dan eerder neigen naar een 400

Reply:
@melsk-r Duidelijk! Dan is mijn vraag inderdaad niet relevant. Desondanks Melvin zijn argument, neig ik toch naar 409 in combinatie met POST. Reden namelijk dat een 400 wordt beschreven als:

" The 400 (Bad Request) status code indicates that the server cannot or
   will not process the request due to something that is perceived to be
   a c**lient error (e.g., malformed request syntax, invalid request
   message framing, or deceptive request routing)**."

Dit impliceert voor mij dat de gebruiker een fout maakt met het bericht. Kijkende naar deze casus is het bericht gewoon goed, maar maakt de gebruiker de fout dat er al een volgindicatie heeft plaatsgevonden. Om die reden zou 409 toch netter zijn en dan met POST aangezien we met PUT dan ook update zouden moeten ondersteunen wat we dus niet willen.

Reply:
besproken in overleg. Wijzigen specs zodat POST maar 1 volgindicatie tegelijk kan zetten (in pull request #19). Dan is foutmelding 409 duidelijk genoeg.